net: lib: coap_client: wait for all acknowledgements

fgervais opened the pull request:
This commit makes sure we continue to wait for extra confirmations even after the request is done so we can handle duplicate confirmations if any.

## Detailed description

rfc7252#section-4.5 specifies that:

"The recipient SHOULD acknowledge each duplicate copy of a
 Confirmable message".

So if, for example, the client sends to a multicast destination address, the server will get multiple requests and will confirm all of them.

Without this commit, the client will set the request to done after receiving the first answer.
From here the request object will be marked as free and the duplicate acknowledgements will stay buffered in the network stack. Once the client tries to send a new request, it will unbuffer those duplicate acknowledgements but now the request object is unallocated so the client won't be able to handle those acknowledgements as duplicates. It will instead treat it as an unexpected ACK.

To work around this issue, rfc7252#section-4.8.2 states that:

"EXCHANGE_LIFETIME is the time from starting to send a Confirmable
 message to the time when an acknowledgement is no longer expected,
 i.e., message-layer information about the message exchange can be
 purged."

Keeping the request object allocated for EXCHANGE_LIFETIME ensures that duplicate acknowledgements can be handled accordingly.

This commit adds a basic implementation of what is stated in the RFC.

EXCHANGE_LIFETIME has been arbitrarily set to 3 * ACK_TIMEOUT which seems more reasonable than the 247 seconds stated in the RFC.

rlubos reviewed (commented):
Looks fine but now the tests fail as the request slots are not freed rightaway. I suggest setting `CONFIG_COAP_INIT_ACK_TIMEOUT_MS ` in the `coap_client` tests to some small lvalue (10 ms?) and adding small delay at the end of the each test case. Actually many tests already have delays at the end, so they might just need some finetuning.

fgervais committed:
tests: coap_client: wait enough for requests to be unallocated

As confirmable requests will stay allocated for (3 * ACK_TIMEOUT), we
need to adjust the timings so all requests are unallocated by the end
of the test.

Signed-off-by: Francois Gervais <francoisgervais@gmail.com>

fgervais committed:
net: lib: coap_client: wait for all acknowledgements

This commit makes sure we continue to wait for extra confirmations even
after the request is done so we can handle duplicate confirmations if any.

Detailed description:

rfc7252#section-4.5 specifies that:

"The recipient SHOULD acknowledge each duplicate copy of a
 Confirmable message".

So if, for example, the client sends to a multicast destination address,
the server will get multiple requests and will confirm all of them.

Without this commit, the client will set the request to done after
receiving the first answer.
From here the request object will be marked as free and the duplicate
acknowledgements will stay buffered in the network stack.
Once the client tries to send a new request, it will unbuffer those
duplicate acknowledgements but now the request object is unallocated
so the client won't be able to handle those acknowledgements as duplicates.
It will instead treat it as an unexpected ACK.

To work around this issue, rfc7252#section-4.8.2 states that:

"EXCHANGE_LIFETIME is the time from starting to send a Confirmable
 message to the time when an acknowledgement is no longer expected,
 i.e., message-layer information about the message exchange can be
 purged."

Keeping the request object allocated for EXCHANGE_LIFETIME ensures that
duplicate acknowledgements can be handled accordingly.

This commit adds a basic implementation of what is stated in the RFC.

EXCHANGE_LIFETIME has been arbitrarily set to 3 * ACK_TIMEOUT which
seems more reasonable than the 247 seconds stated in the RFC.

Signed-off-by: Francois Gervais <francoisgervais@gmail.com>

rlubos reviewed (commented):
One test case is still failing, the following worked for me locally:
```patch
diff --git a/tests/net/lib/coap_client/src/main.c b/tests/net/lib/coap_client/src/main.c
index fd55ba876e2..32089c2adff 100644
--- a/tests/net/lib/coap_client/src/main.c
+++ b/tests/net/lib/coap_client/src/main.c
@@ -405,7 +405,7 @@ ZTEST(coap_client, test_resend_request)
        LOG_INF("Send request");
        ret = coap_client_req(&client, 0, &address, &client_request, NULL);
        zassert_true(ret >= 0, "Sending request failed, %d", ret);
-       k_sleep(K_MSEC(300));
+       k_sleep(K_MSEC(20));
        set_socket_events(ZSOCK_POLLIN);
 
        k_sleep(K_MSEC(100));
```

fgervais commented:
What we currently have seems to match:

```
CONFIG_COAP_INIT_ACK_TIMEOUT_MS + (CONFIG_COAP_INIT_ACK_TIMEOUT_MS / 2)
```

So following this same formula I went with `15 ms`.

All tests passed locally so let see what the automated run says.

jukkar commented on tests/net/lib/coap_client/src/main.c:
```diff
@@ -676,9 +667,8 @@ ZTEST(coap_client, test_unmatching_tokens)
+	k_sleep(K_MSEC(2));
 	clear_socket_events();
-	k_sleep(K_MSEC(500));
+	k_sleep(K_MSEC(700));
```
I have used these kind of sleeps myself in the past, but as the tests run by PR already take long time, these sleeps just add up and make CI runs longer. So it would make sense to try to either lower the sleeps or handle this differently. Anyway, this does not need to be handled in this PR, just a suggestion for future work.
